### DISCUSSION TITLE
How to handle paginated endpoints with no Link header

### DISCUSSION THREAD

**Discussion by bennettscience** [ucfopen/canvasapi#605] [Ideas]
I'm working on #565 right now and have run into an issue with paginating the responses.

[LiveAssessments](https://canvas.instructure.com/doc/api/live_assessments.html) are a tool in Canvas to give feedback on Learning Outcomes in your course. The Magic Marker app used to use this, but Instructure hasn't done anything with it in ages. The problem is in how the endpoint returns paginated responses.

Instead of using a `Link` header that the `PaginatedList` class can use, the links to paginated responses are in the body of the response. This happens for any `live_assessment` endpoint, including for results.

```python
{
  "links": {
    "assessments.results": "http://example.com/courses/1/live_assessments/1/results"
  },
  "meta": {
      "pagination": {
          "per_page": 10,
          "next":"https://www.example.com/api/v1/courses/1/live_assessments/results?page=2&per_page=10"
          "last": "https://example.com/api/v1/courses/1/live_assessments/1/results?page=2&per_page=10",
          "current":"https://example.com/api/v1/courses/1/live_assessments/1/results?page=1&per_page=10",
          "first":"https://example.com/api/v1/courses/1/live_assessments/1/results?page=1&per_page=10",
          "page":1,
          "template":"https://example.com/api/v1/courses/1/live_assessments/1/results?page={page}&per_page=10",
          "count": 20,
          "page_count":2
      }
  }
}
```
There are a couple ways I could handle this:

1. Make the `LiveAssessment` class an iterable on its own. The user could use it just the same as the `PaginatedList` responses.
2. Iterate all of the responses and returning a `list` of live assessments or results, but not make the class itself iterable.
3. Return JSON and note in the docs that the user needs to iterate themselves.

Does one of these stand out as better or worse than another? Given how little it's actually used in Canvas, I don't have a strong opinion one way or another.

**Comment by Thetwam** [comment]
It's been a while since I've looked at the `PaginatedList` code, so this is theoretical...

It might be possible to first try to get the info from `Link` headers, but if none are found, check for a `meta` in the body, before ultimately raising an error.

**Reply by bennettscience** [reply]
Ok, I'll work ok a patch for `PaginatedList` because this would help a couple different endpoints and see if I can get something manageable. #565 is such a niche endpoint, that I don't think it's worth doing a custom implementation here.

**Comment by Thetwam** [comment]
Looks to be resolved in #614
